### PR TITLE
Simple module compilation cache

### PR DESCRIPTION
**Pull request by Mrowqa**

### CODE REVIEW

**Comment by Mrowqa** [comment]
I ran the same test as in https://github.com/CraneStation/wasmtime/pull/190, and the module cache seems to be noticeable slower. I reran the script multiple times. Sometimes, it looks like the compiler can't load the cache (i.e. godot below) - and it's not specific to concrete wasm files. I'll look into this later.
```
$ ./test-cache-efficiency.sh
========
-rw-r--r-- 1 Artur Jamro 197121 13K Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/a.out.wasm
Without cache: 125.911 (100%)
With cache: 107.802 (85.61%)
========
-rw-r--r-- 1 Artur Jamro 197121 31 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/addm1.wasm
Without cache: 66.127 (100%)
With cache: 69.060 (104.43%)
========
-rw-r--r-- 1 Artur Jamro 197121 11M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/AngryBots1.wasm
Without cache: 26158.916 (100%)
With cache: 26601.286 (101.69%)
========
-rw-r--r-- 1 Artur Jamro 197121 51 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/break-to-return.wasm
Without cache: 74.845 (100%)
With cache: 88.894 (118.77%)
========
-rw-r--r-- 1 Artur Jamro 197121 30M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/clang.wasm
Without cache: 34539.732 (100%)
With cache: 1261.255 (3.65%)
========
-rw-r--r-- 1 Artur Jamro 197121 5.7M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/cv-wasm.wasm
Without cache: 12934.096 (100%)
With cache: 318.343 (2.46%)
========
-rw-r--r-- 1 Artur Jamro 197121 306 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/dylib.wasm
Without cache: 60.319 (100%)
With cache: 57.636 (95.55%)
========
-rw-r--r-- 1 Artur Jamro 197121 175 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/fib.wasm
Without cache: 60.571 (100%)
With cache: 46.298 (76.43%)
========
-rw-r--r-- 1 Artur Jamro 197121 35M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/godot.wasm
Without cache: 34057.301 (100%)
With cache: 1075.070 (3.15%)
========
-rw-r--r-- 1 Artur Jamro 197121 36M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/godot-2018-01-18.wasmWithout cache: 34532.884 (100%)
With cache: 35862.678 (103.85%)
========
-rw-r--r-- 1 Artur Jamro 197121 2.8M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/libfaust-wasm.wasm
Without cache: 9853.138 (100%)
With cache: 9301.581 (94.40%)
========
-rw-r--r-- 1 Artur Jamro 197121 12M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/mono.wasm
Without cache: 11042.715 (100%)
With cache: 461.592 (4.18%)
========
-rw-r--r-- 1 Artur Jamro 197121 112K Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/musl.wasm
Without cache: 128.480 (100%)
With cache: 68.865 (53.59%)
========
-rw-r--r-- 1 Artur Jamro 197121 207K Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/oryol-quad.wasm
Without cache: 1418.723 (100%)
With cache: 63.929 (4.50%)
========
-rw-r--r-- 1 Artur Jamro 197121 58K Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/rocket.wasm
Without cache: 133.474 (100%)
With cache: 69.963 (52.41%)
========
-rw-r--r-- 1 Artur Jamro 197121 132 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/s.wasm
Without cache: 69.264 (100%)
With cache: 69.771 (100.73%)
========
-rw-r--r-- 1 Artur Jamro 197121 41 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/simple.wasm
Without cache: 63.018 (100%)
With cache: 76.710 (121.72%)
========
-rw-r--r-- 1 Artur Jamro 197121 1.5M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/sqlite.wasm
Without cache: 6624.709 (100%)
With cache: 121.720 (1.83%)
========
-rw-r--r-- 1 Artur Jamro 197121 54 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/stacky.wasm
Without cache: 74.699 (100%)
With cache: 55.088 (73.74%)
========
-rw-r--r-- 1 Artur Jamro 197121 132 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/stripped.wasm
Without cache: 80.441 (100%)
With cache: 57.893 (71.96%)
========
-rw-r--r-- 1 Artur Jamro 197121 12M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/tanks.wasm
Without cache: 15201.653 (100%)
With cache: 15627.179 (102.79%)
========
-rw-r--r-- 1 Artur Jamro 197121 181 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/test.wasm
Without cache: 89.114 (100%)
With cache: 73.927 (82.95%)
========
-rw-r--r-- 1 Artur Jamro 197121 221 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/trivial.obj.wasm
Without cache: 87.024 (100%)
With cache: 73.491 (84.44%)
========
-rw-r--r-- 1 Artur Jamro 197121 38M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/UE4Game-HTML5-Shipping.wasm
Without cache: 32904.177 (100%)
With cache: 1208.095 (3.67%)
```

**Comment by Mrowqa** [comment]
Note: #204, #205 and #207 already contained in this PR.

**Comment by Mrowqa** [comment]
After fixing the bug, I reran the performance test, adding an extra information how big is the cache and how it compares to the size of input wasm file.
```
$ ./test-cache-efficiency.sh
========
-rw-r--r-- 1 Artur Jamro 197121 13K Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/a.out.wasm
Without cache: 120.975 ms (100%)
With cache: 69.659 ms (57.58%)
Cache size: 56K (428.07%)
========
-rw-r--r-- 1 Artur Jamro 197121 31 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/addm1.wasm
Without cache: 67.011 ms (100%)
With cache: 66.687 ms (99.51%)
Cache size: 1.0K (245.16%)
========
-rw-r--r-- 1 Artur Jamro 197121 11M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/AngryBots1.wasm
Without cache: 26135.619 ms (100%)
With cache: 503.656 ms (1.92%)
Cache size: 65M (636.36%)
========
-rw-r--r-- 1 Artur Jamro 197121 51 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/break-to-return.wasm
Without cache: 75.993 ms (100%)
With cache: 68.509 ms (90.15%)
Cache size: 1.0K (147.05%)
========
-rw-r--r-- 1 Artur Jamro 197121 30M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/clang.wasm
Without cache: 33388.529 ms (100%)
With cache: 1222.714 ms (3.66%)
Cache size: 178M (609.87%)
========
-rw-r--r-- 1 Artur Jamro 197121 5.7M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/cv-wasm.wasm
Without cache: 12418.203 ms (100%)
With cache: 275.034 ms (2.21%)
Cache size: 34M (599.93%)
========
-rw-r--r-- 1 Artur Jamro 197121 306 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/dylib.wasm
Without cache: 73.758 ms (100%)
With cache: 66.064 ms (89.56%)
Cache size: 1.0K (85.94%)
========
-rw-r--r-- 1 Artur Jamro 197121 175 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/fib.wasm
Without cache: 76.186 ms (100%)
With cache: 61.084 ms (80.17%)
Cache size: 1.0K (83.42%)
========
-rw-r--r-- 1 Artur Jamro 197121 35M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/godot.wasm
Without cache: 31443.634 ms (100%)
With cache: 1092.917 ms (3.47%)
Cache size: 146M (426.53%)
========
-rw-r--r-- 1 Artur Jamro 197121 36M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/godot-2018-01-18.wasmWithout cache: 34194.492 ms (100%)
With cache: 1125.242 ms (3.29%)
Cache size: 150M (424.80%)
========
-rw-r--r-- 1 Artur Jamro 197121 2.8M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/libfaust-wasm.wasm
Without cache: 9227.529 ms (100%)
With cache: 178.361 ms (1.93%)
Cache size: 21M (747.86%)
========
-rw-r--r-- 1 Artur Jamro 197121 12M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/mono.wasm
Without cache: 11318.957 ms (100%)
With cache: 472.141 ms (4.17%)
Cache size: 58M (497.20%)
========
-rw-r--r-- 1 Artur Jamro 197121 112K Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/musl.wasm
Without cache: 140.418 ms (100%)
With cache: 67.149 ms (47.82%)
Cache size: 364K (324.13%)
========
-rw-r--r-- 1 Artur Jamro 197121 207K Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/oryol-quad.wasm
Without cache: 1469.463 ms (100%)
With cache: 68.128 ms (4.63%)
Cache size: 1.3M (625.16%)
========
-rw-r--r-- 1 Artur Jamro 197121 58K Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/rocket.wasm
Without cache: 124.815 ms (100%)
With cache: 72.950 ms (58.44%)
Cache size: 260K (451.58%)
========
-rw-r--r-- 1 Artur Jamro 197121 132 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/s.wasm
Without cache: 67.440 ms (100%)
With cache: 59.847 ms (88.74%)
Cache size: 1.0K (282.57%)
========
-rw-r--r-- 1 Artur Jamro 197121 41 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/simple.wasm
Without cache: 69.739 ms (100%)
With cache: 65.552 ms (93.99%)
Cache size: 1.0K (168.29%)
========
-rw-r--r-- 1 Artur Jamro 197121 1.5M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/sqlite.wasm
Without cache: 6574.348 ms (100%)
With cache: 118.789 ms (1.80%)
Cache size: 6.1M (411.17%)
========
-rw-r--r-- 1 Artur Jamro 197121 54 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/stacky.wasm
Without cache: 62.009 ms (100%)
With cache: 68.497 ms (110.46%)
Cache size: 1.0K (138.88%)
========
-rw-r--r-- 1 Artur Jamro 197121 132 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/stripped.wasm
Without cache: 60.417 ms (100%)
With cache: 69.815 ms (115.55%)
Cache size: 1.0K (110.60%)
========
-rw-r--r-- 1 Artur Jamro 197121 12M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/tanks.wasm
Without cache: 20321.763 ms (100%)
With cache: 547.969 ms (2.69%)
Cache size: 72M (602.18%)
========
-rw-r--r-- 1 Artur Jamro 197121 181 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/test.wasm
Without cache: 61.446 ms (100%)
With cache: 70.821 ms (115.25%)
Cache size: 1.0K (180.66%)
========
-rw-r--r-- 1 Artur Jamro 197121 221 Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/trivial.obj.wasm
Without cache: 74.338 ms (100%)
With cache: 56.446 ms (75.93%)
Cache size: 1.0K (95.92%)
========
-rw-r--r-- 1 Artur Jamro 197121 38M Jul 12 14:41 C:/Users/mrowqa/repos/wasm-collection/misc-valid/UE4Game-HTML5-Shipping.wasm
Without cache: 30579.431 ms (100%)
With cache: 1206.165 ms (3.94%)
Cache size: 158M (417.90%)
```

File: wasmtime-environ/src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+            let cache_buf = match bincode::serialize(&data) {
+                Ok(data) => data,
+                Err(err) => {
+                    warn!("Failed to serialize cached code: {}", err);
+                    return;
+                }
+            };
+            match fs::write(p, &cache_buf) {
+                Ok(()) => (),
+                Err(err) => warn!(
```
Comment (sunfishcode) [inline wasmtime-environ/src/cache.rs]: If I understand correctly, if the write fails, `fs::write` may leave a partially-written file behind. We should do an `fs::remove_file` on the path in the error case.

File: wasmtime-environ/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+                #[cfg(windows)]
+                let cache_dir = {
+                    if cache_dir.starts_with("\\\\?\\") {
+                        cache_dir
+                    }
+                    else {
+                        long_path.push(cache_dir.as_os_str());
+                        Path::new(&long_path)
+                    }
+                };
```
Comment (Mrowqa) [inline wasmtime-environ/src/cache.rs]: Just a note: this part is nicely rewritten in #217.

File: wasmtime-environ/src/cache.rs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+use std::ffi::OsString;
+use std::fmt;
+use std::fs;
+#[cfg(windows)]
+use std::path::Path;
+use std::path::PathBuf;
+
+lazy_static! {
+    static ref CACHE_DIR: Option<PathBuf> =
+        match ProjectDirs::from("org", "CraneStation", "wasmtime") {
```
[the remark this answers is not in the document]
Comment (Mrowqa) [reply wasmtime-environ/src/cache.rs]: The docs [say](https://docs.rs/directories/2.0.1/directories/struct.ProjectDirs.html#parameters) it's fine. I'll fix it.

File: wasmtime-environ/src/cache.rs (hunk 1 above)
Comment (Mrowqa) [reply wasmtime-environ/src/cache.rs]: Do you mean calling it anyway? I'm not sure how I should detect this situation from the [Error](https://doc.rust-lang.org/std/io/struct.Error.html). It's good idea to cleanup, so I'll do it, however current code is still correct: next time the file would be read, not parsed correctly, and overwritten with new cached data.

**Comment by Mrowqa** [comment]
Another thought: we probably should have some smoke tests: for the `cache_conf` and for some basic module caching - assuring we can successfully load the cache and it would contain the original data. Should I write them and add to this PR, create a new PR with them, or postpone it for now?

**Comment by sunfishcode** [comment]
I think it's fine to add the tests in followup PRs in this case.

**Comment by Mrowqa** [comment]
Ok. First bigger change merged - it feels great :)

**Pull request merged**